File: src/main/java/ru/ezhov/rocket/action/types/GroupRocketActionUi.java
package ru.ezhov.rocket.action.types;

import ru.ezhov.rocket.action.RocketActionUiRepository;
import ru.ezhov.rocket.action.api.RocketActionConfigurationProperty;
import ru.ezhov.rocket.action.api.RocketActionSettings;
import ru.ezhov.rocket.action.api.RocketActionUi;
import ru.ezhov.rocket.action.icon.AppIcon;
import ru.ezhov.rocket.action.icon.IconRepositoryFactory;
import ru.ezhov.rocket.action.infrastructure.ReflectionRocketActionUiRepository;
import ru.ezhov.rocket.action.icon.IconService;

import javax.swing.ImageIcon;
import javax.swing.JMenu;
import javax.swing.SwingWorker;
import java.awt.Component;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.ExecutionException;

public class GroupRocketActionUi extends AbstractRocketAction {

    private static final String TYPE = "GROUP";
    private static final String LABEL = "label";
    private static final String ICON_URL = "iconUrl";
    private static final String DESCRIPTION = "description";

    @Override
    public Component create(RocketActionSettings settings) {
        JMenu menu = new JMenu(ConfigurationUtil.getValue(settings.settings(), LABEL));
        menu.setIcon(new ImageIcon(this.getClass().getResource("/load_16x16.gif")));
        menu.setToolTipText(ConfigurationUtil.getValue(settings.settings(), DESCRIPTION));

        new GroupSwingWorker(menu, settings).execute();

        return menu;
    }

    @Override
    public String type() {
        return TYPE;
    }

    @Override
    public String description() {
        return "description";
    }

    @Override
    public List<RocketActionConfigurationProperty> properties() {
        return Arrays.asList(
                createRocketActionProperty(LABEL, "TEST", true),
                createRocketActionProperty(DESCRIPTION, "TEST", true),
                createRocketActionProperty(ICON_URL, "URL for icon", false)
        );
    }

    private static class GroupSwingWorker extends SwingWorker<List<Component>, String> {
        private JMenu parentMenu;
        private RocketActionSettings settings;

        public GroupSwingWorker(JMenu menu, RocketActionSettings settings) {
            this.parentMenu = menu;
            this.settings = settings;
        }

        @Override
        protected List<Component> doInBackground() throws Exception {
            RocketActionUiRepository rocketActionUiRepository = new ReflectionRocketActionUiRepository(); //TODO: сделать нормально
            return createGroup(rocketActionUiRepository, settings.actions(), parentMenu);
        }

        private List<Component> createGroup(RocketActionUiRepository rocketActionUiRepository, List<RocketActionSettings> actionSettings, JMenu parent) {
            List<Component> children = new ArrayList<>();
            for (RocketActionSettings settings : actionSettings) {
                if (settings.type().equals(TYPE)) {
                    JMenu menu = new JMenu(ConfigurationUtil.getValue(settings.settings(), LABEL));
                    menu.setIcon(new ImageIcon(this.getClass().getResource("/load_16x16.gif")));
                    menu.setToolTipText(ConfigurationUtil.getValue(settings.settings(), DESCRIPTION));

                    new GroupSwingWorker(menu, settings).execute();

                    createGroup(rocketActionUiRepository, settings.actions(), menu);
                    children.add(menu);
                } else {
                    final Optional<RocketActionUi> actionUi = rocketActionUiRepository.by(settings.type());
                    actionUi.ifPresent(rocketActionUi -> children.add(rocketActionUi.create(settings)));
                }
            }

            return children;
        }

        @Override
        protected void done() {
            try {
                final List<Component> components = this.get();
                components.forEach(c -> parentMenu.add(c));
            } catch (InterruptedException e) {
                e.printStackTrace();
            } catch (ExecutionException e) {
                e.printStackTrace();
            }

            parentMenu.setIcon(
                    IconService.load(
                            Optional.ofNullable(settings.settings().get(ICON_URL)),
                            IconRepositoryFactory.instance().by(AppIcon.PROJECT)
                    )
            );
        }
    }
}

File: src/main/kotlin/ru/ezhov/rocket/action/types/GroupRocketActionUi.kt
package ru.ezhov.rocket.action.types

import ru.ezhov.rocket.action.RocketActionUiRepository
import ru.ezhov.rocket.action.api.RocketActionConfigurationProperty
import ru.ezhov.rocket.action.api.RocketActionSettings
import ru.ezhov.rocket.action.api.RocketActionUi
import ru.ezhov.rocket.action.icon.AppIcon
import ru.ezhov.rocket.action.icon.IconRepositoryFactory
import ru.ezhov.rocket.action.icon.IconService
import ru.ezhov.rocket.action.infrastructure.ReflectionRocketActionUiRepository
import java.awt.Component
import java.util.concurrent.ExecutionException
import java.util.function.Consumer
import javax.swing.ImageIcon
import javax.swing.JMenu
import javax.swing.SwingWorker

class GroupRocketActionUi : AbstractRocketAction() {
    override fun create(settings: RocketActionSettings): Component {
        val menu = JMenu(ConfigurationUtil.getValue(settings.settings(), LABEL))
        menu.icon = ImageIcon(this.javaClass.getResource("/load_16x16.gif"))
        menu.toolTipText = ConfigurationUtil.getValue(settings.settings(), DESCRIPTION)
        GroupSwingWorker(menu, settings).execute()
        return menu
    }

    override fun type(): String {
        return TYPE
    }

    override fun description(): String {
        return "description"
    }

    override fun properties(): List<RocketActionConfigurationProperty> {
        return listOf(
                createRocketActionProperty(LABEL, LABEL, "TEST", true),
                createRocketActionProperty(DESCRIPTION, DESCRIPTION, "TEST", true),
                createRocketActionProperty(ICON_URL, ICON_URL, "URL for icon", false)
        )
    }

    override fun name(): String = "Группа"

    private class GroupSwingWorker(private val parentMenu: JMenu, private val settings: RocketActionSettings) : SwingWorker<List<Component?>, String?>() {
        @Throws(Exception::class)
        override fun doInBackground(): List<Component?> {
            val rocketActionUiRepository: RocketActionUiRepository = ReflectionRocketActionUiRepository() //TODO: сделать нормально
            rocketActionUiRepository.load()
            return createGroup(rocketActionUiRepository, settings.actions(), parentMenu)
        }

        private fun createGroup(rocketActionUiRepository: RocketActionUiRepository, actionSettings: List<RocketActionSettings?>?, parent: JMenu): List<Component?> {
            val children: MutableList<Component?> = ArrayList()
            for (settings in actionSettings!!) {
                if (settings!!.type() == TYPE) {
                    val menu = JMenu(ConfigurationUtil.getValue(settings.settings(), LABEL))
                    menu.icon = ImageIcon(this.javaClass.getResource("/load_16x16.gif"))
                    menu.toolTipText = ConfigurationUtil.getValue(settings.settings(), DESCRIPTION)
                    GroupSwingWorker(menu, settings).execute()
                    createGroup(rocketActionUiRepository, settings.actions(), menu)
                    children.add(menu)
                } else {
                    rocketActionUiRepository.by(settings.type())
                            ?.let { rocketActionUi: RocketActionUi ->
                                children.add(rocketActionUi.create(settings))
                            }
                }
            }
            return children
        }

        override fun done() {
            try {
                val components = this.get()
                components!!.forEach(Consumer { c: Component? -> parentMenu.add(c) })
            } catch (e: InterruptedException) {
                e.printStackTrace()
            } catch (e: ExecutionException) {
                e.printStackTrace()
            }
            parentMenu.icon = IconService().load(
                    settings.settings()[ICON_URL].orEmpty(),
                    IconRepositoryFactory.repository.by(AppIcon.PROJECT)
            )
        }
    }

    companion object {
        private const val TYPE = "GROUP"
        private const val LABEL = "label"
        private const val ICON_URL = "iconUrl"
        private const val DESCRIPTION = "description"
    }
}